daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/TypeCalculator.scala

```diff
@@ -0,0 +1,453 @@
+   *     error by throwing
+   * 2) DPathCompiledExpressions provides a top level interface to initiate evaluation which accepts a ParseOrUnparseState,
+   *      and a second interface to evaluate subexpressions which accepts a DState
+   */
+
+  def inputTypeCalc(x: A, xType: NodeInfo.Kind): (Maybe[B], Maybe[Error])
+  def outputTypeCalc(x: B, xType: NodeInfo.Kind): (Maybe[A], Maybe[Error])
+
+  def inputTypeCalcParse(pstate: PState, context: RuntimeData, x: A, xType: NodeInfo.Kind): Maybe[B] = {
+    val (ans, err) = inputTypeCalc(x, xType)
+    Assert.invariant(ans.isDefined ^ err.isDefined)
+
+    if (err.isDefined) {
+      val diag = new ParseError(Maybe(context.schemaFileLocation), Maybe(pstate.currentLocation), err.get)
+      pstate.setFailed(diag)
+    }
+
+    //In event of an error, we still want to return Maybe.Nope, which happens
+    //to be what ans would have
+    ans
+  }
+  def outputTypeCalcUnparse(ustate: UState, context: RuntimeData, x: B, xType: NodeInfo.Kind): Maybe[A] = {
+    val (ans, err) = outputTypeCalc(x, xType)
+    Assert.invariant(ans.isDefined ^ err.isDefined)
+
+    if (err.isDefined) {
+      val diag = new ParseError(Maybe(context.schemaFileLocation), Maybe(ustate.currentLocation), err.get)
+      ustate.setFailed(diag)
+    }
+
+    //In event of an error, we still want to return Maybe.Nope, which happens
+    //to be what ans would have
+    ans
+  }
+
+  /*
+   * It appears that dpath expressions actually throw errors.
+   * See the definition of DAFError
+   */
+  def inputTypeCalcRun(dstate: DState, x: A, xType: NodeInfo.Kind): Unit = {
+    val context = dstate.runtimeData.get
+    val (ans, err) = inputTypeCalc(x, xType)
+    Assert.invariant(ans.isDefined ^ err.isDefined)
+
+    if (err.isDefined) {
+      val diag = new ParseError(Maybe(context.schemaFileLocation), dstate.contextLocation, err.get)
+      throw diag
+      //      throw FNErrorFunctionException(Maybe(context.schemaFileLocation), dstate.contextLocation, err.get)
+    }
+
+    dstate.setCurrentValue(ans)
+
+  }
+  def outputTypeCalcRun(dstate: DState, x: B, xType: NodeInfo.Kind): Unit = {
+    val context = dstate.runtimeData.get
+    val (ans, err) = outputTypeCalc(x, xType)
+    Assert.invariant(ans.isDefined ^ err.isDefined)
+
+    if (err.isDefined) {
+      val diag = new ParseError(Maybe(context.schemaFileLocation), dstate.contextLocation, err.get)
+      throw diag
+    }
+
+    dstate.setCurrentValue(ans)
+
+  }
+
+  def supportsParse: Boolean = true
+  def supportsUnparse: Boolean = true
+}
+
+class KeysetValueTypeCalculatorOrdered[A <: AnyRef, B <: AnyRef](valueMap: HashMap[A, B], rangeTable: Seq[(RangeBound[A], RangeBound[A], B)], unparseMap: HashMap[B, A],
+                                                                 srcType: NodeInfo.Kind, dstType: NodeInfo.Kind)
+  extends TypeCalculator[A, B](srcType, dstType) {
+
+  override def inputTypeCalc(x: A, xType: NodeInfo.Kind): (Maybe[B], Maybe[Error]) = {
```

Comment:
It this is going to allocate a Tuple, where _1 is success and _2 is an error, we should just use an Either. Allocation is going to happen either way, so we don't gain anything by using two maybes.

Though, I'm not even sure the error message return is necessary. The error message only contains the value of ``x` and whenther it was "key" or "value"`, which the all should be able to know. I would argue this should just return a a Maybe[B], and the caller can figure out the appropriate error message if Maybe.Nope is returned.